File: MODEL/object_1d.cpp
#include "object_1d.h"

Object_1D::Object_1D() : Space(), length(1) {}

Object_1D::Object_1D(int l) : Space(), length(l) {}

Object_1D::Object_1D(int l, int risol) : Space(risol), length(l) {}

Object_1D::Object_1D(int l, int risol, RGBHex c) : Space(risol, c), length(l) {}

int Object_1D::CalcolaSuperficie()
{
    return 0; //poichè un punto o una linea non possegono superficie
}

Status Object_1D::getStatus() const
{
    Status s = Space::getStatus();
    //sistemare lunghezza
    s.l = this->getLength();
    s.lCm = this->getLCm();
    s.lInch = this->getLInch();
    return s;
}

int Object_1D::getLength() const
{
    return length;
}

void Object_1D::setLength(int l)
{
    length = l;
}

Object_1D& Object_1D::operator+(const Space &x)
{
    const Object_1D* a = dynamic_cast<const Object_1D*>(&x);
    if(a)
    {
        int val = this->getLength() + a->getLength();
        this->setLength(val);
        this->setRisoluzione(a->getRisoluzione()); //la somma ritorna semplicemente la risoluzione del parametro passato
        return *this;
    }
    //non fa niente in questo caso se non è del tipo giusto
    throw Ecc_Object1D_somma_ObjectDiverso();
}



Object_1D& Object_1D::operator-(const Space &x)
{
    const Object_1D* a = dynamic_cast<const Object_1D*>(&x);
    if(a)
    {
        int val = this->getLength() - a->getLength();
        if(val <=0) val = 1;
        this->setLength(val);
        this->setRisoluzione(a->getRisoluzione()); //la sottrazione ritorna semplicemente la risoluzione del parametro passato
        return *this;
    }
    throw Ecc_Object1D_sottrazione_ObjectDiverso();
}

Object_1D& Object_1D::operator*(const Space &x) //questo operatore viene chiamato con oggetto di inocazione di tipo Object1d quindi siamo sicuri che quello sarà il suo tipo, ritorniamo uno space
{
    const Object_1D* a = dynamic_cast<const Object_1D*>(&x);
    if(a)
    {
        int val = this->getLength() * a->getLength();
        this->setLength(val);
        this->setRisoluzione(a->getRisoluzione()); //la moltiplicazione ritorna semplicemente la risoluzione del parametro passato
        return *this;
    }
    throw Ecc_Object1D_moltiplicazione_ObjectDiverso();

}



Object_1D& Object_1D::operator/(const Space &x)
{    
    const Object_1D* a = dynamic_cast<const Object_1D*>(&x);
    if(a)
    {
        int val = this->getLength() / a->getLength();
        if(val <=0) val = 1;
        this->setLength(val);
        this->setRisoluzione(a->getRisoluzione()); //la moltiplicazione ritorna semplicemente la risoluzione del parametro passato
        return *this;
    }

    throw Ecc_Object1D_divisione_Object1D();
}

Object_1D& Object_1D::operator+(const RGBHex &x)
{
    RGBHex b;
    int val = x.getColorInt() + this->col.getColorInt();
    if (val>16777215) val=16777215;
    b.setColorInt(val);
    this->setColor(b);
    return *this;
}

Object_1D& Object_1D::operator-(const RGBHex &x)
{
    RGBHex b;
    int val = col.getColorInt() - x.getColorInt();
    if(val<0) val=0;
    b.setColorInt(val);
    this->setColor(b);
    return *this;
}

Object_1D& Object_1D::operator*(const RGBHex &x)
{
    this->setColor(x);
    return *this;
}

Object_1D& Object_1D::operator/(const RGBHex &x)
{
    col.getColorInt() == x.getColorInt() ? this->setColor(RGBHex("000000")) : this->setColor(col);
    return *this;


}

void Object_1D::setLCm(double l)
{
    length = static_cast<int> (Space::dpi*(l/2.54)); // l/2.54 serve per convertire la lunghezza l da cm a pollici
    if ((Space::dpi*(l/2.54))-length > 0) length++; // Arrotondo per eccesso
}

double Object_1D::getLCm() const
{
    double a = static_cast<double>(length);
    double b = static_cast<double>(Space::dpi);
    return ((a/b)*2.54);
}

void Object_1D::setLInch(double l)
{
    length = static_cast<int> (Space::dpi*l);
    if ((Space::dpi*l)-length > 0) length++; // Arrotondo per eccesso
}

double Object_1D::getLInch() const
{
    double a = static_cast<double>(length);
    double b = static_cast<double>(length);
    return (a/b);
}

File: MODEL-JAVA/src/Kalk/Object_1D.java
package Kalk;

public class Object_1D extends Space {
	private int length;
	
	//Costruttori
	public Object_1D() {
		super();
		length=1;
	}
	
	public Object_1D(int l) {
		super();
		length=l;
	}
	
	public Object_1D(int l, int risol) {
		super(risol);
		length=l;	
	}
	
	public Object_1D(int l, int risol, RGBHex c) {
		super(risol, c);
		length=l;
	}
	
	public Status getStatus() 
	{
	    Status s = super.getStatus();
	    //sistemare lunghezza
	    s.l = this.getLength();
	    s.lCm = this.getLCm();
	    s.lInch = this.getLInch();
	    return s;
	}

	private double getLInch() {

	    double a = (double)length;
	    double b = (double)length;
	    return (a/b);
	}

	private double getLCm() {
	    double a = (double)length;
	    double b = (double)dpi;
	    return ((a/b)*2.54);
	}

	int getLength() {
		return length;
	}
	

	void setLength(int l)
		{
			length = l;
		}

	void setLInch(double l) {
    		length = (int) (dpi*l);
    		if ((dpi*l)-length > 0) length++; // Arrotondo per eccesso
		}
	
	void setLCm(double l) {
	    length = (int) (dpi*(l/2.54)); // l/2.54 serve per convertire la lunghezza l da cm a pollici
	    if ((dpi*(l/2.54))-length > 0) length++; // Arrotondo per eccesso
	}


	//Object_1D con RGBHex
	
	@Override
	Object_1D sum(RGBHex x)  {
	    RGBHex b = new RGBHex();
	    int val = x.getColorInt() + this.col.getColorInt();
	    if (val>16777215) val=16777215;
	    b.setColorInt(val);
	    this.setColor(b);
	    
	    return this;
	}

	@Override
	Object_1D sub(RGBHex x) {
	    RGBHex b = new RGBHex();
	    int val = col.getColorInt() - x.getColorInt();
	    if(val<0) val=0;
	    b.setColorInt(val);
	    this.setColor(b);
	    return this;
	}

	@Override
	Object_1D mult(RGBHex x) {
	    this.setColor(col);
	    return this;
	}

	@Override
	Object_1D div(RGBHex x) {
	    if(col.getColorInt() == (x.getColorInt()))
	    this.setColor(new RGBHex("000000"));
	    else
	    	this.setColor(col);
	    return this;
	}
	
	//Object_1D con Space

	@Override
	Object_1D sum(Space x) throws Exceptions{
	    if(x instanceof Object_1D)
	    {
	    	Object_1D a = (Object_1D) x;
	        int val = this.getLength() + a.getLength();
	        this.setLength(val);
	        this.setRisoluzione(a.getRisoluzione()); //la somma ritorna semplicemente la risoluzione del parametro passato
	        super.col = super.col.sum( x.col);
	        return this;
	    }
	    //non fa niente in questo caso se non è del tipo giusto
	    throw new Exceptions("Object1D non può sommarsi con tipi diversi di space da esso");
	}

	@Override
	Object_1D sub(Space x) throws Exceptions{
	    if(x instanceof Object_1D)
	    {
	    	Object_1D a = (Object_1D) x;
	        int val = this.getLength() - a.getLength();
	        if(val <=0) val = 1;
	        this.setLength(val);
	        this.setRisoluzione(a.getRisoluzione()); //la sottrazione ritorna semplicemente la risoluzione del parametro passato
	        super.col = super.col.sub( x.col);
	        return this;
	    }
	    throw new Exceptions("Object1D non può sottrarsi con tipi diversi di space da esso");
	}

	@Override
	Object_1D mult(Space x) throws Exceptions {
	    if(x instanceof Object_1D)
	    {
	    	Object_1D a = (Object_1D) x;
	        int val = this.getLength() * a.getLength();
	        this.setLength(val);
	        this.setRisoluzione(a.getRisoluzione()); //la moltiplicazione ritorna semplicemente la risoluzione del parametro passato
	        super.col = super.col.sum( x.col);
	        return this;
	    }
	    throw new Exceptions("Object1D non può moltiplicarsi con tipi diversi di space da esso");
	}

	@Override
	Object_1D div(Space x) throws Exceptions {
	    if(x instanceof Object_1D)
	    {
	    	Object_1D a = (Object_1D) x;
	        int val = this.getLength() / a.getLength();
	        if(val <=0) val = 1;
	        this.setLength(val);
	        this.setRisoluzione(a.getRisoluzione()); //la moltiplicazione ritorna semplicemente la risoluzione del parametro passato
	        super.col = super.col.sub( x.col);
	        return this;
	    }

	    throw new Exceptions("Object1D non può dividersi con tipi diversi di space da esso");
	}

}
